core/src/impl/Kokkos_Profiling.hpp

```diff
@@ -134,6 +136,323 @@ void resume_tools();
+ *
+ * If you're tuning a reducer, it might be a complex or a simple reducer
+ * (an example of simple would be one where the join is just "+".
+ *
+ * Unfortunately these two paths are very different in terms of which classes
+ * get instantiated. Thankfully, all of this complexity is encoded in the
+ * ReducerType. If it's a "simple" reducer, this will be Kokkos::InvalidType,
+ * otherwise it'll be something else.
+ *
+ * If the type is complex, for the code to be generally right you _must_
+ * pass an instance of that ReducerType to functions that determine
+ * eligible team sizes. If the type is simple, you can't construct one,
+ * you use the simpler 2-arg formulation of team_size_recommended/max.
+ *
+ *
+ */
+
+namespace Impl {
+
+struct SimpleTeamSizeCalculator {
+  template <typename Policy, typename Functor, typename Tag>
+  int get_max_team_size(const Policy& policy, const Functor& functor,
+                        const Tag& tag) {
+    auto max = policy.team_size_max(functor, tag);
+    return max;
+  }
+};
+
+// when we have a complex reducer, we need to pass an
+// instance to team_size_recommended/max. Reducers
+// aren't default constructible, but they are
+// constructible from a reference to an
+// instance of their value_type so we construct
+// a value_type and temporary reducer here
+template <typename ReducerType>
+struct ComplexReducerSizeCalculator {
+  template <typename Policy, typename Functor, typename Tag>
+  int get_max_team_size(const Policy& policy, const Functor& functor,
+                        const Tag& tag) {
+    using value_type = typename ReducerType::value_type;
+    value_type value;
+    ReducerType reducer_example = ReducerType(value);
+    return policy.team_size_max(functor, reducer_example, tag);
+  }
+};
+
+}  // namespace Impl
+
+template <class Functor, class TagType, class... Properties>
+void tune_policy(const size_t /**tuning_context*/, const std::string& label_in,
+                 Kokkos::TeamPolicy<Properties...>& policy,
+                 const Functor& functor, const TagType& tag) {
+  if (policy.auto_team_size() || policy.auto_vector_length()) {
+    std::string label = label_in;
+    if (label_in.empty()) {
+      using policy_type =
+          typename std::remove_reference<decltype(policy)>::type;
+      using work_tag = typename policy_type::work_tag;
+      Kokkos::Impl::ParallelConstructName<Functor, work_tag> name(label);
+      label = name.get();
+    }
+    if (team_tuners.find(label) == team_tuners.end()) {
+      team_tuners.emplace(label, Kokkos::Tools::Experimental::TeamSizeTuner(
+                                     label, policy, functor, tag,
+                                     Impl::SimpleTeamSizeCalculator{}));
+    }
+    auto& tuner = team_tuners[label];
+    tuner.tune(policy);
+    std::cout << "Policy deets for " << label << ": " << policy.team_size()
+              << ", " << policy.vector_length() << std::endl;
```

Comment:
Do you still want this `cout`? The other constructor does not have it.

Reply:
Nope, that's a pretty bad miss on my part